services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -103,17 +115,68 @@ public String encode(Token token) {
     }
 
     @Override
-    public <T> T decodeClientJWT(String token, ClientModel client, Class<T> clazz) {
-        if (token == null) {
+    public <T> T decodeClientJWT(String jwt, ClientModel client, BiConsumer<JOSE, ClientModel> jwtValidator, Class<T> clazz) {
+        if (jwt == null) {
             return null;
         }
-        try {
-            JWSInput jws = new JWSInput(token);
 
-            String signatureAlgorithm = jws.getHeader().getAlgorithm().name();
+        JOSE joseToken = JOSEParser.parse(jwt);
+
+        jwtValidator.accept(joseToken, client);
+
+        if (joseToken instanceof JWE) {
+            try {
+                Optional<KeyWrapper> activeKey;
+                String kid = joseToken.getHeader().getKeyId();
+                Stream<KeyWrapper> keys = session.keys().getKeysStream(session.getContext().getRealm());
+
+                if (kid == null) {
+                    activeKey = keys.filter(k -> KeyUse.ENC.equals(k.getUse()) && k.getPublicKey() != null)
+                            .sorted(Comparator.comparingLong(KeyWrapper::getProviderPriority).reversed())
+                            .findFirst();
+                } else {
+                    activeKey = keys.filter(k -> k.getKid().equals(kid)).findAny();
+                }
+
+                JWE jwe = JWE.class.cast(joseToken);
+                Key privateKey = activeKey.map(KeyWrapper::getPrivateKey)
+                        .orElseThrow(() -> new RuntimeException("Could not find private key for decrypting token"));
+
+                jwe.getKeyStorage().setDecryptionKey(privateKey);
+
+                byte[] content = jwe.verifyAndDecodeJwe().getContent();
 
+                try {
+                    JOSE jws = JOSEParser.parse(new String(content));
+
+                    if (jws instanceof JWSInput) {
+                        jwtValidator.accept(jws, client);
+                        return verifyJWS(client, clazz, (JWSInput) jws);
+                    }
+                } catch (Exception ignore) {
+                    // try to decrypt content as is
+                }
+
+                return JsonSerialization.readValue(content, clazz);
+            } catch (IOException cause) {
+                throw new RuntimeException("Failed to deserialize JWT", cause);
+            } catch (JWEException cause) {
+                throw new RuntimeException("Failed to decrypt JWT", cause);
+            }
+        }
+
+        return verifyJWS(client, clazz, (JWSInput) joseToken);
+    }
+
+    private <T> T verifyJWS(ClientModel client, Class<T> clazz, JWSInput jws) {
+        try {
+            String signatureAlgorithm = jws.getHeader().getAlgorithm().name();
             ClientSignatureVerifierProvider signatureProvider = session.getProvider(ClientSignatureVerifierProvider.class, signatureAlgorithm);
+
             if (signatureProvider == null) {
+                if (jws.getHeader().getAlgorithm().equals(org.keycloak.jose.jws.Algorithm.none)) {
```

Comment:
It seems that until now, the "none" algorithm was not supported by the method `decodeClientJWT`, but now it is supported with this change. My personal vote is to not support "none" algorithm here, but rather support it at the caller level (EG. AuthzEndpointRequestObjectParser) if it wants to.

IMO the support for "none" algorithm can be the reason of various security bugs, so it will be nice to be more careful and make it at least slightly harder to avoid introducing security issues.

Or at least, if you really prefer to support "none" algorithm at DefaultTokenManager.decodeClientJWT, add the javadoc to the method `TokenManager.decodeClientJWT` to mention that `none` algorithm is supported by this method.

Reply:
I'm not sure if this applies here because there is a validator that calling code can use to make sure the JWT, mainly headers, are in conformance to their requirements.

For instance, the authorization endpoint parser checks whether the algorithm is defined and matches the one defined within the client config.

Reply:
You're right. Some other points might not have proper validation for any. So changed the validator to fail by default when none.